### PR TITLE
Apache Ignite Dataset

### PR DESCRIPTION
**Pull request by dmitrievanthony**
This is a proposal to add `IgniteDataset` that allows to work with [Apache Ignite](https://ignite.apache.org/). 

Apache Ignite is a memory-centric distributed database, caching, and processing platform for transactional, analytical, and streaming workloads, delivering in-memory speeds at petabyte scale. This proposal is a part of a more global initiative to implement so called "TensorFlow on Apache Ignite" ([IGNITE-8335](https://issues.apache.org/jira/browse/IGNITE-8335), [Design Document](https://docs.google.com/document/d/1jROIahK1rc7bSgOvhJhfpMqIGvht_IE8zn5NAt6x8ks)).

The integration is based on Apache Ignite [Binary Client Protocol](https://apacheignite.readme.io/v2.6/docs/binary-client-protocol) and TensorFlow [tf.data.Dataset](https://www.tensorflow.org/guide/datasets). More information about supported features you can find in [README.md](https://github.com/dmitrievanthony/tensorflow/blob/apache-ignite-dataset/tensorflow/contrib/ignite/README.md) of this module.

Tests have also been added. They use docker to hide configuration complexity, so that the implemented functionality can be tested quite simply by manual run.

### CODE REVIEW

File: tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "ignite_binary_object_parser.h"
+#include "ignite_dataset.h"
+
+#ifndef IGNITE_CLIENT_H
```
Comment (mrry) [inline tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h]: For header guards like this one, please use the full path (and this one seems not to match the filename):

```c++
#ifndef TENSORFLOW_CONTRIB_IGNITE_KERNELS_IGNITE_DATASET_ITERATOR_H_`
```

File: tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "ignite_dataset.h"
+#include <stdlib.h>
+#include "tensorflow/core/framework/dataset.h"
+
+namespace tensorflow {
+
```
Comment (mrry) [inline tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc]: Please wrap this class and the registration in an anonymous namespace.

**Review by mrry — CHANGES_REQUESTED** [review]
Thanks for resubmitting this PR. Since the PR is extra-large, it may take several weeks to review in its entirety. I've taken a quick pass and made some comments on style that apply to all of the code, so please apply those first.

**Comment by dmitrievanthony** [comment]
Thank you for reviewing, @mrry. Really hope it won't take more than two weeks to review and make this request merge-ready because in Apache Ignite we have big plans about it. Feel free to contact me any time if you have questions about this PR.

I've updated the code in accordance with all comments you have so far, so I would be grateful if you could continue your review.

**Comment by dmitrievanthony** [comment]
Hi, @mrry, @martinwicke. Any updates?

By the way, could you please run CI tests so that I am able to fix now if something wrong?

**Comment by dmitrievanthony** [comment]
Guys, @mrry, @martinwicke, I really understand that you are under pressure of all these pull requests, but it's been more than 10 days since your last comment here, so let me ping you and ask to take a look.

**Comment by martinwicke** [comment]
I've retriggered the tests. Sorry for the wait.

**Comment by dmitrievanthony** [comment]
Hi, @martinwicke. Thank you for restarting tests. Sorry for bothering, but I still can't fully reproduce CI build.

I see that during the last run only one required test failed (one `pylint` check to be precise). I have just fixed it, so could you please restart tests again?

And one more question. Should only `required` tests be green or all? I see that a lot of "none-required" tests fail in absolutely another modules with timeout and other types of exceptions, so I don't feel like I can fix them.

**Comment by martinwicke** [comment]
@mrry can you take a look at the changes?

**Comment by dmitrievanthony** [comment]
Hi, @mrry, @martinwicke. I feel yourself a bit lost here. It's been 17 days since last @mrry comment and the PR is on the 3rd page already. Guys, could you please pay attention to this request?

**Comment by perfinion** [comment]
@dmitrievanthony sorry for all the delay. I re-triggered the tests, looks like there is some issue with the windows build and clang-format (The XLA tests fail a lot, I'm not sure they are related), can you take a look?

Also can you rebase and fix up the commits? The Merge Commits in the PR are weird i dont follow them. If you do an interactive rebase on origin/master you can drop them and fix up the commits.

I'm not really familiar with these areas of TF so can't review all of it but I'll do what I can.

File: tensorflow/contrib/ignite/__init__.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+   https://apacheignite.readme.io/v2.6/docs/binary-client-protocol.
+
+@@IgniteDataset
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.contrib.ignite.python.ops.ignite_dataset_ops \
```
[the remark this answers is not in the document]
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/__init__.py]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "ignite_binary_object_parser.h"
+
+namespace ignite {
```
[the remark this answers is not in the document]
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+  uint8_t object_type_id = *ptr;
+  ptr += 1;
+
+  switch (object_type_id) {
+    case BYTE: {
+      tensorflow::Tensor tensor(tensorflow::cpu_allocator(),
+                                tensorflow::DT_UINT8, {});
+      tensor.scalar<tensorflow::uint8>()() = *((uint8_t*)ptr);
+      ptr += 1;
+      out_tensors.emplace_back(std::move(tensor));
```
[the remark this answers is not in the document]
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "ignite_binary_object_parser.h"
+
+namespace ignite {
+
+tensorflow::Status BinaryObjectParser::Parse(
+    uint8_t*& ptr, std::vector<tensorflow::Tensor>& out_tensors,
```
[the remark this answers is not in the document]
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_binary_object_parser.cc]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_client.cc [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+==============================================================================*/
+
+#ifndef IGNITE_CLIENT_H
+#define IGNITE_CLIENT_H
+#include "ignite_client.h"
+#endif
+
+namespace ignite {
+
+tensorflow::Status Client::ReadByte(uint8_t& data) {
```
[the remark this answers is not in the document]
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_client.cc]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_dataset.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      const override;
+  tensorflow::string DebugString() const override;
+
+ protected:
+  tensorflow::Status AsGraphDefInternal(
+      tensorflow::SerializationContext* ctx, DatasetGraphDefBuilder* b,
+      tensorflow::Node** output) const override;
+
+ private:
+  const std::string cache_name;
```
[the remark this answers is not in the document]
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_dataset.h]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+#include "ignite_plain_client.h"
+#include "ignite_ssl_wrapper.h"
+#include "tensorflow/core/platform/logging.h"
+
+#include <time.h>
+#include <chrono>
+
+namespace ignite {
+
+#define CHECK_STATUS(status) \
```
[the remark this answers is not in the document]
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.cc]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h (hunk 1 above)
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_dataset_iterator.h]: Fixed.

File: tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc (hunk 2 above)
Comment (dmitrievanthony) [reply tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc]: Fixed.

**Comment by dmitrievanthony** [comment]
Hi @perfinion. I've fixed code style and removed "merge requests", but as result I've recreated the branch, so could you please continue with #22210?

Regarding builds on Windows I see the same issues I had and asked here: https://groups.google.com/a/tensorflow.org/forum/#!topic/build/ePYss0Kxcu4. It looks like we need to add `copt=-DWIN32_LEAN_AND_MEAN` on CI server for Windows builds. What do you think?

**Comment by perfinion** [comment]
oh, you could have just force-pushed to this branch name. but sure, I've run tests and re-assigned that PR instead.

**Pull request closed without merging**